flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaShuffleProducer.java

```diff
@@ -0,0 +1,194 @@
+	 * @param next element to handle
+	 * @param context context needed to handle the element
+	 * @throws FlinkKafkaException
+	 */
+	@Override
+	public void invoke(KafkaTransactionState transaction, IN next, Context context) throws FlinkKafkaException {
+		checkErroneous();
+
+		// write timestamp to Kafka if timestamp is available
+		Long timestamp = context.timestamp();
+
+		int[] partitions = getPartitions(transaction);
+		int partitionIndex;
+		try {
+			partitionIndex = KeyGroupRangeAssignment
+				.assignKeyToParallelOperator(keySelector.getKey(next), partitions.length, partitions.length);
+		} catch (Exception e) {
+			throw new RuntimeException("Fail to assign a partition number to record");
+		}
+
+		ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
+			defaultTopicId, partitionIndex, timestamp, null, kafkaSerializer.serializeRecord(next, timestamp));
+		pendingRecords.incrementAndGet();
+		transaction.producer.send(record, callback);
+	}
+
+	/**
+	 * This is the function invoked to handle each watermark.
+	 * @param transaction transaction state;
+	 *                    watermark are written to Kafka (if needed) in transactions
+	 * @param watermark watermark to handle
+	 * @throws FlinkKafkaException
+	 */
+	@Override
+	public void invoke(KafkaTransactionState transaction, Watermark watermark) throws FlinkKafkaException {
+		checkErroneous();
+
+		int[] partitions = getPartitions(transaction);
+		int subtask = getRuntimeContext().getIndexOfThisSubtask();
+
+		// broadcast watermark
+		for (int partition : partitions) {
+			ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(
+				defaultTopicId, partition, watermark.getTimestamp(), null, kafkaSerializer.serializeWatermark(watermark, subtask));
+			pendingRecords.incrementAndGet();
+			transaction.producer.send(record, callback);
+		}
+	}
+
+	private int[] getPartitions(KafkaTransactionState transaction) {
+		int[] partitions = topicPartitionsMap.get(defaultTopicId);
+		if (partitions == null) {
+			partitions = getPartitionsByTopic(defaultTopicId, transaction.producer);
+			topicPartitionsMap.put(defaultTopicId, partitions);
+		}
+
+		Preconditions.checkArgument(partitions.length == numberOfPartitions);
+
+		return partitions;
+	}
+
+	/**
+	 * Flink Kafka Shuffle Serializer.
+	 */
+	public static final class KafkaSerializer<IN> implements Serializable {
+		public static final int TAG_REC_WITH_TIMESTAMP = 0;
+		public static final int TAG_REC_WITHOUT_TIMESTAMP = 1;
+		public static final int TAG_WATERMARK = 2;
+
+		private final TypeSerializer<IN> serializer;
+
+		private transient DataOutputSerializer dos;
+
+		KafkaSerializer(TypeSerializer<IN> serializer) {
+			this.serializer = serializer;
+		}
+
+		/**
+		 * Format: TAG, (timestamp), record.
+		 */
+		byte[] serializeRecord(IN record, Long timestamp) {
+			if (dos == null) {
+				dos = new DataOutputSerializer(16);
+			}
+
+			try {
+				if (timestamp == null) {
+					dos.writeInt(TAG_REC_WITHOUT_TIMESTAMP);
```

Comment:
My gut feeling is that we need to forward also some events, such as end of partition. 
I'm also not entirely sure how to deal with checkpoint barriers (see general remark).
In any case, we should use `StreamElementSerializer` instead of implementing it again.

Reply:
Mark as resolved since the entire API is re-organized.

StreamElementSerializer is discussed offline and summarized in the following comments.